lalrpop/src/lr1/lane_table/construct/group.rs

```diff
@@ -0,0 +1,165 @@
+        if let Some(context_set) = context_set_opt {
+            match self.context_set_mutref(group).inplace_union(context_set) {
+                Err(_) => {
+                    debug!("Group::merge_state: failed to merge group={:?} with state={:?}", group, state);
+                    false
+                } 
+                Ok(_) => {
+                    debug!("Group::merge_state: successfull merge group={:?} with state={:?} context_set={:?}", group, state, self.context_set_ref(group));
+                    true
+                } 
+            }
+        } else {
+            true
+        }
+    }
+
+    fn unify_key(&self, group: Group) -> GroupUnifyKey {
+        self.unification_keys[group.index]
+    }
+
+    fn context_set(&mut self, group: Group) -> ContextSet {
+        let key = self.unify_key(group);
+        let empty = ContextSet::new(0);
+        replace(&mut self.context_sets[self.unification_table.probe_value(key).index], empty)
+    }
+
+    fn context_set_mutref(&mut self, group: Group) -> &mut ContextSet {
+        let key = self.unify_key(group);
+        &mut self.context_sets[self.unification_table.probe_value(key).index]
+    }
+
+    pub fn context_set_ref(&mut self, group: Group) -> &ContextSet {
+        let key = self.unify_key(group);
+        &self.context_sets[self.unification_table.probe_value(key).index]
+    }
+
+    pub fn context_set_ref_for_state(&mut self, state: StateIndex) -> &ContextSet {
+        let group = self.groups[state.0].unwrap().clone();
+        self.context_set_ref(group)
+    }
+
+    pub fn unify_groups(&mut self, group1: Group, group2: Group) {
+        let key1 = self.unify_key(group1);
+        let key2 = self.unify_key(group2);
+        assert!(self.unification_table.unify_var_var(key1, key2).is_ok())
+    }
+
+    pub fn merge_groups(&mut self, group1: Group, group2: Group) -> bool {
+        let context_set = {
+            let mut context_set1 = self.context_set(group1);
+            let context_set2 = self.context_set(group2);
+            debug!("Groups::merge_groups: group={:?} context_set{:?}", group1, context_set1);
+            debug!("Groups::merge_groups: group={:?} context_set{:?}", group2, context_set2);
+            match context_set1.inplace_union(&context_set2) {
+                Ok(_) => context_set1,
+                Err(_) => {
+                    // rollback
+                    self.context_sets[group1.index] = context_set1;
+                    self.context_sets[group2.index] = context_set2;
+                    debug!("Groups::merge_grops: merging groups {:?} and {:?} failed", group1, group2);
+                    return false
+                }
+            }
+        };
+
+        self.unify_groups(group1, group2);
+        debug!("Groups::merge_groups: successfull merge of groups {:?} and {:?} context_set={:?}", group1, group2, context_set);
+        self.context_sets[max(group1, group2).index] = context_set;
```

Comment:
Hmm, instead of using `max`, couldn't we query to find the new root of `group1` (or `group2`)? All things being equal, it's usually better if you let the UF algorithm pick the new root on its own, so it can balance the trees.